src/semaphore.rs

```diff
@@ -349,6 +358,19 @@ impl Drop for SemaphoreGuard<'_> {
 #[derive(Debug)]
 pub struct SemaphoreGuardArc(Arc<Semaphore>);
 
+impl SemaphoreGuardArc {
+    /// Drops the guard _without_ releasing the acquired permit.
+    /// (Will still decrement the `Arc` reference count.)
+    #[inline]
+    pub fn forget(self) {
+        let mut manual = core::mem::ManuallyDrop::new(self);
+
+        // Drop the inner `Arc` in order to decrement the reference count.
+        // SAFETY: `manual` not used after this
+        let _arc = unsafe { addr_of_mut!(manual.0).read() };
```

Comment:
I'd rather avoid the `unsafe` here. I would prefer to make the type a wrapper around `Option<Arc<...>>`, and then make this just `take` the `Arc` out without any of the other dropping.

Reply:
For both cases, I value safety over precedent here. `smol-rs` doesn't have the resources to conduct thorough `unsafe` audits (outside of off-the-shelf tools like Miri), so the less unsafe code we have the better.

Reply:
> For both cases

What do you mean by "both cases" here? Are you saying I should remove the `unsafe` in the `RwLock` impl also? (The performance cost would be much greater over there compared to here.)

Reply:
Not in this PR, but I am skeptical of premature optimizations like this.

Reply:
I've removed the `unsafe`.

Reply:
>  (The performance cost would be much greater over there compared to here.)

Specifically, it would mean a branch on every access to the value behind the `RwLock`.

Reply:
We can benchmark this in the future